innovation_sweet_spots/analysis/prototyping/public_discourse_analysis/pd_data_collection_utils.py

```diff
@@ -0,0 +1,270 @@
+    return articles_by_year
+        
+
+def extract_text_from_html(html, tags: Iterator[str]) -> Iterator[str]:
+    """
+    Extract text from specified html tags.
+
+    Parameters
+    ----------
+    html (str): content of an article in html format.
+    tags (list): list of tags to extract content from.
+
+    Returns
+    -------
+    no_html (list): list of chunks (e.g. paragraphs) of text extracted from the tags.
+
+    """
+    segments = BeautifulSoup(html, "html.parser").find_all(tags)
+    no_html = [seg.get_text() for seg in segments]
+    return no_html
+
+
+def get_text_segments(articles: Iterator[dict], tags: Iterator[str]) -> Iterator[str]:
+    """
+    Extract segments of article text from Guardian articles in html format.
+
+    Parameters
+    ----------
+    articles (list): list of dicts containing result of the Guardian
+    search_content function.
+    tags (list): list of tags to extract content from.
+
+    Returns
+    -------
+    clean_segments (list): nested list of article segments.
+
+    """
+    article_html = [article["fields"]["body"] for article in articles]
+    clean_segments = [extract_text_from_html(article, tags) for article in article_html]
+    return clean_segments
+
+
+def get_article_text(articles: Iterator[dict], tags: Iterator[str]):
+    """
+    Extract full text of an article from Guardian articles in html format.
+
+    Parameters
+    ----------
+    articles (list): list of dicts containing result of the Guardian
+    search_content function.
+    tags (list): list of tags to extract content from.
+
+    Returns
+    -------
+    article_text (list): list of article text.
+
+    """
+    article_segments = get_text_segments(articles, tags)
+    article_text = [" ".join(segment) for segment in article_segments]
+    return article_text
+
+
+def get_article_metadata(grouped_articles, fields_to_extract=["id"]):
+    """
+    Extract useful article fields from raw data returned by the Guardian API.
+
+    Parameters
+    ----------
+    grouped_articles (dict): articles grouped by year using itertools.
+    fields_to_extract (list): list of fields. the default is "id", other useful
+                                fields are "webUrl" and "webTitle"
+
+    Returns
+    -------
+    year_article_df_combined (pandas.core.frame.DataFrame): dataframe with
+    content of article fields.
+
+    """
+    year_article_dfs = []
+    for year, articles in grouped_articles.items():
+        year_data = dict()
+        for field in fields_to_extract:
+            article_field = [article[field] for article in articles]
+            year_data[field] = article_field
+            year_data["year"] = year
+        year_article_df = pd.DataFrame(year_data)
+        year_article_dfs.append(year_article_df)
+    year_article_df_combined = pd.concat(year_article_dfs)
+    year_article_df_combined = year_article_df_combined.drop_duplicates()
+    return year_article_df_combined
```

Comment:
Is there another reason to group the articles by year other than to pass them into this function? If there is, then I would rewrite the function in a slightly neater way, e.g. by passing in a set of records directly into a `DataFrame` like this

```suggestion
records = []
for year, articles in grouped_articles.items():
    for article in articles:
        article['year'] = year
        records.append(article)
articles_df = (pd.DataFrame(records)
               .drop_duplicates())
return articles_df
```

However, we're basically just unpacking the grouping here, which makes me wonder whether you could just pass in the raw articles here as an argument and remove `sort_by_year` altogether.

Reply:
Re-worked to use raw articles in get_article_metadata and get_article_text_df.